xopt/base.py

```
@@ -171,8 +170,9 @@ def validate_data(cls, v, info: ValidationInfo):
         # also add data to generator
         # TODO: find a more robust way of doing this
         generator = info.data["generator"]
-        if not isinstance(generator, SequentialGenerator):
-            # sequential generators must maintain their own state
+
+        # Some generators need to maintain their own state (such as sequential generators)
+        if not isinstance(generator, StateOwner):
```

see comment in xopt generator file

---

I checked out the other classes that use `.set_data` and `NelderMeadGenerator._set_data` seems to be doing some validation as well. `ExtremumSeekingGenerator._set_data` only takes the last row of the data? Then `RCDSGenerator` doesn't implement `._set_data` which means it should raise an exception when reloaded. Then for the generators that aren't sequential generators I'm not sure how not calling .add_data would change things.

It looks like just doing the assignment will change a lot of behavior so maybe it's a good refactor for its own PR? Maybe someone else can advise on how to handle this best?

---

yes, we can leave this to a future PR, we will fix the RCDS loading issue shortly